### PR TITLE
Use gRPC interceptor instead of data converter for encoding/decoding

### LINKED ISSUE #367: [Bug] Ensure remote codec used for all payloads in and out
### Describe the bug

When setting `--codec-endpoint` it appears to set a custom data converter globally for a few uses, but we need to make sure it's used for all input and ouput payloads (that are not search attributes)

### PR DESCRIPTION
**Pull request by dandavison**
Fixes #367

Depends on unmerged PR https://github.com/temporalio/sdk-go/pull/1303

### What was changed
- Do payload encoding/decoding for all client requests, using a gRPC interceptor
- Stop doing payload encoding/decoding in the dataconverter
- Eliminate the global mutable dataconverter model

### Why?
- `temporal workflow show` was respecting `--codec-endpoint` but `temporal workflow show --output json` was not

### Why do we think this change is correct?
For this change to maintain previous semantics, all the code paths that were using the custom dataconverter must have been operating on data that was either a response obtained using `SDKClient`, `FrontendClient`, or `OperatorClient`, or is being prepared for a request via one of those clients. I have added comments inline establishing that this is true.

For this change to be correct, we require that all code paths that send/receive data containing `Payloads` are using `SDKClient`, `FrontendClient`, or `OperatorClient`.

### How was this tested
- Manual testing only so far, using https://github.com/temporalio/samples-python/tree/main/encryption: <table><tr><td>image</td></tr></table>

- [ ] More testing needed

### Checklist
1. Addresses #367 w.r.t. `workflow show --output json`
2. Any docs updates needed?
No

### CODE REVIEW

**Comment by cretz** [comment]
@dandavison - Look at https://pkg.go.dev/go.temporal.io/api/proxy#VisitPayloads for visiting all payloads (be sure to exclude search attributes). https://github.com/temporalio/sdk-go/blob/v1.25.1/converter/grpc_interceptor.go#L47 is an example of it being used.

File: schedule/schedule_commands.go [hunk 1]
```diff
@@ -459,7 +459,7 @@ func DescribeSchedule(c *cli.Context) error {
 	if sw := s.Action.GetStartWorkflow(); sw != nil {
 		item.StartWorkflow = sw
 		item.WorkflowType = sw.WorkflowType.GetName()
-		item.Input = dataconverter.CustomDataConverter().ToStrings(sw.Input)
+		item.Input = converter.GetDefaultDataConverter().ToStrings(sw.Input)
```
Comment (dandavison) [inline schedule/schedule_commands.go:462]: Correctness: `s` is `resp.Schedule`, and `resp` was obtained using `frontendClient`, https://github.com/temporalio/cli/blob/80a17b4de02405e0edc646754b7d48094bf51095/schedule/schedule_commands.go#L412
and `frontendClient` is created in a way that has the gRPC interceptor set: https://github.com/temporalio/cli/blob/80a17b4de02405e0edc646754b7d48094bf51095/schedule/schedule_commands.go#L35

File: schedule/schedule_commands.go [hunk 2]
```diff
@@ -479,13 +479,13 @@ func DescribeSchedule(c *cli.Context) error {
 	if fields := resp.Memo.GetFields(); len(fields) > 0 {
 		item.Memo = make(map[string]string, len(fields))
 		for k, payload := range fields {
-			item.Memo[k] = dataconverter.CustomDataConverter().ToString(payload)
+			item.Memo[k] = converter.GetDefaultDataConverter().ToString(payload)
```
Comment (dandavison) [inline schedule/schedule_commands.go:482]: Correctness: using the same `resp` as [above](https://github.com/temporalio/cli/pull/384/files?show-deleted-files=true&show-viewed-files=true&file-filters%5B%5D=#r1406065941).

File: schedule/schedule_commands.go [hunk 3]
```diff
@@ -479,13 +479,13 @@ func DescribeSchedule(c *cli.Context) error {
...
 		}
 	}
 	if fields := resp.SearchAttributes.GetIndexedFields(); len(fields) > 0 {
 		item.SearchAttributes = make(map[string]string, len(fields))
 		for k, payload := range fields {
-			item.SearchAttributes[k] = dataconverter.DefaultDataConverter().ToString(payload)
+			item.SearchAttributes[k] = converter.GetDefaultDataConverter().ToString(payload)
```
Comment (dandavison) [inline schedule/schedule_commands.go:488]: Correctness: same `resp` as [above](https://github.com/temporalio/cli/pull/384/files?show-deleted-files=true&show-viewed-files=true&file-filters%5B%5D=#r1406065941).

File: schedule/schedule_commands.go [hunk 4]
```diff
@@ -656,7 +656,7 @@ func encodeMemo(memo map[string]interface{}) (*commonpb.Memo, error) {
 	if len(memo) == 0 {
 		return nil, nil
 	}
-	dc := dataconverter.CustomDataConverter()
+	dc := converter.GetDefaultDataConverter()
```
Comment (dandavison) [inline schedule/schedule_commands.go:659]: Correctness: here we are in `encodeMemo()`. This helper function which is used only in `getMemoAndSearchAttributesForSchedule` which is used only in `CreateSchedule`: https://github.com/temporalio/cli/blob/80a17b4de02405e0edc646754b7d48094bf51095/schedule/schedule_commands.go#L230
 
There we are creating a request, which will be sent using `frontendClient`, which has the gRPC interceptor configured: https://github.com/temporalio/cli/blob/80a17b4de02405e0edc646754b7d48094bf51095/schedule/schedule_commands.go#L219

File: schedule/schedule_commands.go [hunk 5]
```diff
@@ -672,7 +672,7 @@ func encodeSearchAttributes(sa map[string]interface{}) (*commonpb.SearchAttribut
 	if len(sa) == 0 {
 		return nil, nil
 	}
-	dc := dataconverter.DefaultDataConverter()
+	dc := converter.GetDefaultDataConverter()
```
Comment (dandavison) [inline schedule/schedule_commands.go:675]: Correctness: only used in `getMemoAndSearchAttributesForSchedule`, therefore correct by [above](https://github.com/temporalio/cli/pull/384/files?show-deleted-files=true&show-viewed-files=true&file-filters%5B%5D=#r1406077577).

File: workflow/workflow_commands.go [hunk 6]
```diff
@@ -595,7 +595,7 @@ func queryWorkflowHelper(c *cli.Context, queryType string) error {
 	if queryResponse.QueryRejected != nil {
 		fmt.Printf("Query was rejected, workflow has status: %v\n", queryResponse.QueryRejected.GetStatus())
 	} else {
-		queryResult := stringify.AnyToString(queryResponse.QueryResult, true, 0, dataconverter.CustomDataConverter())
+		queryResult := stringify.AnyToString(queryResponse.QueryResult, true, 0, converter.GetDefaultDataConverter())
```
Comment (dandavison) [inline workflow/workflow_commands.go:598]: Correctness: `queryResponse` was obtained using the frontend client, which was created using the usual factory, which sets the gRPC interceptor: https://github.com/temporalio/cli/blob/80a17b4de02405e0edc646754b7d48094bf51095/workflow/workflow_commands.go#L550

File: workflow/workflow_commands.go [hunk 7]
```diff
@@ -781,7 +781,7 @@ func convertDescribeWorkflowExecutionResponse(c *cli.Context, resp *workflowserv
 		}
 
 		if pendingActivity.GetHeartbeatDetails() != nil {
-			pendingActivityStr.HeartbeatDetails = stringify.AnyToString(pendingActivity.GetHeartbeatDetails(), true, 0, dataconverter.CustomDataConverter())
+			pendingActivityStr.HeartbeatDetails = stringify.AnyToString(pendingActivity.GetHeartbeatDetails(), true, 0, converter.GetDefaultDataConverter())
```
Comment (dandavison) [inline workflow/workflow_commands.go:784]: Correctness, `pendingActivity` derives from a response which was passed into this function by `DescribeWorkflow`, which used the standard frontend client factory:

https://github.com/temporalio/cli/blob/80a17b4de02405e0edc646754b7d48094bf51095/workflow/workflow_commands.go#L687

File: workflow/workflow_commands.go [hunk 8]
```diff
@@ -845,7 +845,7 @@ func printRunStatus(c *cli.Context, event *historypb.HistoryEvent) {
 		fmt.Printf("  Retry status: %s\n", event.GetWorkflowExecutionTimedOutEventAttributes().GetRetryState())
 	case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_CANCELED:
 		fmt.Printf("  Status: %s\n", color.Red(c, "CANCELED"))
-		details := stringify.AnyToString(event.GetWorkflowExecutionCanceledEventAttributes().GetDetails(), true, 0, dataconverter.CustomDataConverter())
+		details := stringify.AnyToString(event.GetWorkflowExecutionCanceledEventAttributes().GetDetails(), true, 0, converter.GetDefaultDataConverter())
```
Comment (dandavison) [inline workflow/workflow_commands.go:848]: Correctness: same argument as [above](https://github.com/temporalio/cli/pull/384/files?show-deleted-files=true&show-viewed-files=true&file-filters%5B%5D=#r1406087982).

File: workflow/workflow_commands.go [hunk 9]
```diff
@@ -1577,7 +1577,7 @@ func findWorkflowStatusValue(name string) (enumspb.WorkflowExecutionStatus, bool
 // historyEventToString convert HistoryEvent to string
 func historyEventToString(e *historypb.HistoryEvent, printFully bool, maxFieldLength int) string {
 	data := getEventAttributes(e)
-	return stringify.AnyToString(data, printFully, maxFieldLength, dataconverter.CustomDataConverter())
+	return stringify.AnyToString(data, printFully, maxFieldLength, converter.GetDefaultDataConverter())
```
Comment (dandavison) [inline workflow/workflow_commands.go:1580]: Correctness: the event being formatted was obtained via the history table iterator https://github.com/temporalio/cli/blob/80a17b4de02405e0edc646754b7d48094bf51095/workflow/workflow_commands.go#L273 

The only place the history table iterator is constructed is in `printWorkflowProgress`, which we have [already established](https://github.com/temporalio/cli/pull/384/files?show-deleted-files=true&show-viewed-files=true&file-filters%5B%5D=#r1406087982) uses the gRPC interceptor.

File: go.mod [hunk 10]
```diff
@@ -157,3 +156,5 @@ require (
 )
 
 replace github.com/grpc-ecosystem/grpc-gateway => github.com/temporalio/grpc-gateway v1.17.0
+
+replace go.temporal.io/sdk => github.com/dandavison/temporalio-sdk-go v1.26.0
```
Comment (Quinn-With-Two-Ns) [inline go.mod:160]: remove?

Comment (Quinn-With-Two-Ns) [reply go.mod:160]: Oh I see your depending on unmerged go changes

Comment (dandavison) [reply go.mod]: Yep that's right.

**Comment by Quinn-With-Two-Ns** [comment]
@dandavison did you test search attributes? Those don't appear to be excluded by `NewPayloadCodecGRPCClientInterceptor `?

**Review by Quinn-With-Two-Ns — APPROVED** [review]
LGTM!

File: go.mod [hunk 11]
```diff
@@ -157,3 +156,5 @@ require (
 )
 
 replace github.com/grpc-ecosystem/grpc-gateway => github.com/temporalio/grpc-gateway v1.17.0
+
+replace go.temporal.io/sdk => github.com/dandavison/temporalio-sdk-go v1.30.0
```
Comment (cretz) [inline go.mod:160]: Does this PR need to be updated for #420 and have this removed?

Comment (dandavison) [reply go.mod:160]: I'll close this one and open a PR adding the release candidate cherry pick to main.

**Comment by dandavison** [comment]
Superseded by https://github.com/temporalio/cli/pull/421

**Pull request closed without merging**